### PR TITLE
[PM-9875] Server configurations 

### PR DESCRIPTION
**Pull request by andrebispo5**
## 🎟️ Tracking

[PM-9875 -  Implement fetching the /config endpoint](https://bitwarden.atlassian.net/browse/PM-9875)
## 📔 Objective

Implement fetching and saving the server data retrieved from the /config endpoint, it will contain the server version and feature flags state.

This endpoint should be called:
- On Account Switch
- On Environment Switch

## 📸 Screenshots

https://github.com/user-attachments/assets/15011ab8-96ef-40d3-9a36-36206bf10302

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed
  issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: app/src/test/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryTest.kt [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        configService = configService,
+        dispatcherManager = fakeDispatcherManager
+    )
+
+    @BeforeEach
+    fun setUp() {
+    }
+
+    @AfterEach
+    fun tearDown() {
```
Comment (david-livefront) [inline app/src/test/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryTest.kt:40]: If we don't need these, we can just remove them

File: app/src/main/java/com/x8bit/bitwarden/data/platform/datasource/disk/model/ServerConfig.kt [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+import java.time.ZonedDateTime
+
+/**
+ * A higher-level wrapper around [ConfigResponseJson] that provides a timestamp
+ * to check if a sync is necessary
+ *
+ * @property lastSync The [ZonedDateTime] of the last sync.
+ * @property serverData The raw [ConfigResponseJson] that contains specific data of the
+ * server configuration
+ *
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/platform/datasource/disk/model/ServerConfig.kt:14]: Wanna drop the extra line here

File: app/src/main/java/com/x8bit/bitwarden/data/platform/datasource/disk/ConfigDiskSourceImpl.kt [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    private val json: Json,
+) : BaseDiskSource(sharedPreferences = sharedPreferences),
+    ConfigDiskSource {
+
+    override var serverConfig: ServerConfig?
+        get() = getString(key = SERVER_CONFIGURATIONS)?.let { json.decodeFromStringOrNull(it) }
+        set(value) {
+            putString(
+                key = SERVER_CONFIGURATIONS,
+                value = value?.let { json.encodeToString(value) },
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/platform/datasource/disk/ConfigDiskSourceImpl.kt:28]: Don't use the `value` inside the `let`, just use `it`

File: app/src/main/java/com/x8bit/bitwarden/data/platform/repository/EnvironmentRepositoryImpl.kt [hunk 4]
```diff
@@ -30,6 +32,10 @@ class EnvironmentRepositoryImpl(
             .toEnvironmentUrlsOrDefault()
         set(value) {
             environmentDiskSource.preAuthEnvironmentUrlData = value.environmentUrlData
+            scope.launch {
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/platform/repository/EnvironmentRepositoryImpl.kt:35]: Firing off a network request is a bit of a heavy side effect to have directly in the setter. A better option would be to subscribe to the `environmentStateFlow`, this could be done in the `init` block of the `ServerConfigRepositoryImpl`.

File: app/src/main/java/com/x8bit/bitwarden/data/platform/repository/EnvironmentRepositoryImpl.kt [hunk 5]
```diff
@@ -51,7 +57,11 @@ class EnvironmentRepositoryImpl(
                     ?.activeAccount
                     ?.settings
                     ?.environmentUrlData
-                    ?.let { environmentDiskSource.preAuthEnvironmentUrlData = it }
+                    ?.let {
+                        environmentDiskSource.preAuthEnvironmentUrlData = it
+                        /// Fetch new server configs on active account  change
+                        serverConfigRepository.getServerConfig(forceRefresh = true)
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/platform/repository/EnvironmentRepositoryImpl.kt:63]: This should not be needed if you refresh based on the `environmentStateFlow`

File: app/src/main/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryImpl.kt [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+                onSuccess = { configResponse ->
+                    val serverConfig = ServerConfig(
+                        lastSync = Instant.now().toEpochMilli(),
+                        serverData = configResponse
+                    )
+                    configDiskSource.serverConfig = serverConfig
+                    return serverConfig
+                },
+                onFailure = {
+                    throw it
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryImpl.kt:42]: This seems a bit weird to me? We just crash if the network request fails?

File: app/src/main/java/com/x8bit/bitwarden/data/platform/datasource/disk/ConfigDiskSourceImpl.kt (hunk 3 above)
Comment (andrebispo5) [reply app/src/main/java/com/x8bit/bitwarden/data/platform/datasource/disk/ConfigDiskSourceImpl.kt:28]: Fixed it on `EnvironmentDiskSourceImpl` as well

File: app/src/main/java/com/x8bit/bitwarden/data/platform/repository/EnvironmentRepositoryImpl.kt [hunk 7]
```diff
@@ -51,7 +57,11 @@ class EnvironmentRepositoryImpl(
                     ?.activeAccount
                     ?.settings
                     ?.environmentUrlData
-                    ?.let { environmentDiskSource.preAuthEnvironmentUrlData = it }
+                    ?.let {
+                        environmentDiskSource.preAuthEnvironmentUrlData = it
+                        // Fetch new server configs on active account  change
```
Comment (shannon-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/platform/repository/EnvironmentRepositoryImpl.kt:62]:
```suggestion
                        // Fetch new server configs on active account change
```

**Review by shannon-livefront — COMMENTED** [review]
The code looks good! I have a few generic questions:
1. Is there a way to test that this is working in the app?
2. What work is left to do to fully integrate feature flags?
3. Can we wrap some of this logic into the `BitwardenFeatureFlagManager` so that the rest of the app can have a clean interface to use to access the feature flag states?

File: app/src/main/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryImpl.kt (hunk 6 above)
Comment (andrebispo5) [reply app/src/main/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryImpl.kt:42]: Forgot about it. Nope, we should just handle the success and return the state if the request failed.

**Comment by andrebispo5** [comment]
> The code looks good! I have a few generic questions:
> 
> 1. Is there a way to test that this is working in the app?
> 2. What work is left to do to fully integrate feature flags?
> 3. Can we wrap some of this logic into the `BitwardenFeatureFlagManager` so that the rest of the app can have a clean interface to use to access the feature flag states?

1. So far we are only able to test it by looking into the requests made by the app using the debugger.
2. The remaining work is to expand `BitwardenFeatureFlagManager` with interfaces to consume the flag data provided by `ServerConfigRepository`. [Task for this body of work](https://bitwarden.atlassian.net/browse/PM-9401).
3. The logic to manage the feature flags local and server will live in the `BitwardenFeatureFlagManager`. The manager will use the `ServerConfigRepository` to get the server flag data.

**Comment by andrebispo5** [comment]
@shannon-livefront Added video of the endpoint being executed on account change and environment change. 👍

**Review by shannon-livefront — APPROVED** [review]

File: app/src/main/java/com/x8bit/bitwarden/data/platform/repository/EnvironmentRepositoryImpl.kt [hunk 8]
```diff
@@ -51,7 +51,9 @@ class EnvironmentRepositoryImpl(
                     ?.activeAccount
                     ?.settings
                     ?.environmentUrlData
-                    ?.let { environmentDiskSource.preAuthEnvironmentUrlData = it }
+                    ?.let {
+                        environmentDiskSource.preAuthEnvironmentUrlData = it
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/platform/repository/EnvironmentRepositoryImpl.kt:55]: Why is this change here?

File: app/src/main/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryImpl.kt [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+import java.time.Clock
+import java.time.Instant
+
+/**
+ * Primary implementation of [ServerConfigRepositoryImpl].
+ */
+class ServerConfigRepositoryImpl(
+    private val configDiskSource: ConfigDiskSource,
+    private val configService: ConfigService,
+    private val clock: Clock = Clock.systemDefaultZone(),
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryImpl.kt:22]: You should not need this default value, it should just be injectable.

File: app/src/main/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryImpl.kt [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+        if (needsRefresh || forceRefresh) {
+            configService.getConfig().onSuccess { configResponse ->
+                val serverConfig = ServerConfig(
+                    lastSync = clock.instant().toEpochMilli(),
+                    serverData = configResponse,
+                )
+                configDiskSource.serverConfig = serverConfig
+                return serverConfig
+            }
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryImpl.kt:54]: Are we just ignoring the error then?

File: app/src/main/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryImpl.kt [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+        }
+
+        // If we are unable to retrieve a configuration from the server,
+        // fall back to the local configuration.
+        return localConfig
+    }
+
+    override val serverConfigStateFlow: StateFlow<ServerConfig?>
+        get() = configDiskSource.serverConfigFlow
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryImpl.kt:63]: Can we use chain formatting here:

```kotlin
    get() = configDiskSource
        .serverConfigFlow
        .stateIn(
```

File: app/src/test/java/com/x8bit/bitwarden/data/platform/datasource/disk/ConfigDiskSourceTest.kt [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+import kotlinx.serialization.json.Json
+import org.junit.jupiter.api.Assertions.assertEquals
+import org.junit.jupiter.api.Assertions.assertNull
+import org.junit.jupiter.api.Test
+import java.time.Instant
+
+class ConfigDiskSourceTest {
+    private val fakeSharedPreferences = FakeSharedPreferences()
+
+    private val json = Json {
```
Comment (david-livefront) [inline app/src/test/java/com/x8bit/bitwarden/data/platform/datasource/disk/ConfigDiskSourceTest.kt:20]: You can just do this:
```kotlin
private val json = PlatformNetworkModule.providesJson()
```

That will ensure that you are always testing with identical configuration.

File: app/src/test/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryTest.kt [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+            getConfig()
+        } returns CONFIG_RESPONSE_JSON.asSuccess()
+    }
+    private val environmentRepository = FakeEnvironmentRepository().apply {
+        environment = Environment.Us
+    }
+
+    @BeforeEach
+    fun setUp() {
+        fakeConfigDiskSource.serverConfig = null
```
Comment (david-livefront) [inline app/src/test/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryTest.kt:39]: Why is this needed?

File: app/src/test/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryTest.kt [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+    }
+
+    private val repository = ServerConfigRepositoryImpl(
+        configDiskSource = fakeConfigDiskSource,
+        configService = configService,
+        environmentRepository = environmentRepository,
+        dispatcherManager = fakeDispatcherManager,
```
Comment (david-livefront) [inline app/src/test/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryTest.kt:46]: Make sure to pass in a fixed clock so your tests based on time always work

File: app/src/test/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryTest.kt [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+        repository.serverConfigStateFlow.test {
+            val newConfig = awaitItem()
+            assertNotEquals(
+                serverConfigBefore,
+                newConfig,
+            )
+
+            assertEquals(
+                SERVER_CONFIG.serverData,
+                newConfig!!.serverData,
```
Comment (david-livefront) [inline app/src/test/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryTest.kt:65]: Can we just validate the entire `newConfig`, instead of comparing a partial update.

File: app/src/test/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryTest.kt [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+
+    @Test
+    fun `getServerConfig should fetch a new server configuration with force refresh as true`() =
+        runTest {
+            coEvery {
+                configService.getConfig()
+            } returns CONFIG_RESPONSE_JSON.copy(version = "NEW VERSION").asSuccess()
+
+            val testConfig = SERVER_CONFIG.copy(
+                lastSync = Instant.now().toEpochMilli(),
```
Comment (david-livefront) [inline app/src/test/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryTest.kt:78]: Please use the `Clock`

File: app/src/main/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryImpl.kt (hunk 10 above)
Comment (andrebispo5) [reply app/src/main/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryImpl.kt:54]: Yes and returning the current saved value. This call will run quite frequently, so it will just try again later.

File: app/src/test/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryTest.kt (hunk 13 above)
Comment (andrebispo5) [reply app/src/test/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryTest.kt:39]: It was needed for 2 of the tests but I can move it to them

File: app/src/test/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryTest.kt (hunk 15 above)
Comment (andrebispo5) [reply app/src/test/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryTest.kt:65]: With the fixed clock I believe now we can, let me change that.

File: app/src/test/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryTest.kt (hunk 13 above)
Comment (david-livefront) [reply app/src/test/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryTest.kt:39]: Isn't the default `null` though?

File: app/src/test/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryTest.kt [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+        repository.serverConfigStateFlow.test {
+            val newConfig = awaitItem()
+            assertNotNull(
+                newConfig,
+            )
+
+            assertEquals(
+                SERVER_CONFIG,
+                newConfig,
+            )
```
Comment (david-livefront) [inline app/src/test/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryTest.kt:77]: We can probably simplify this:

```kotlin
repository.serverConfigStateFlow.test {
    assertEquals(
        SERVER_CONFIG,
        awaitItem(),
    )
}
```

File: app/src/test/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryTest.kt [hunk 18]
```diff
@@ -0,0 +1,199 @@
...
+            coEvery {
+                configService.getConfig()
+            } returns CONFIG_RESPONSE_JSON.copy(version = "NEW VERSION").asSuccess()
+
+            fakeConfigDiskSource.serverConfig = SERVER_CONFIG.copy(
+                lastSync = fixedClock.instant().toEpochMilli(),
+            )
+
+            assertNotNull(
+                fakeConfigDiskSource.serverConfig,
```
Comment (david-livefront) [inline app/src/test/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryTest.kt:93]: We shouldn't need all these `assertNotNull`s since the `assertEquals` will catch all that

File: app/src/test/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryTest.kt [hunk 19]
```diff
@@ -0,0 +1,199 @@
...
+    fun `getServerConfig should fetch a new server configuration if there is none in state`() =
+        runTest {
+            assertNull(
+                fakeConfigDiskSource.serverConfig,
+            )
+
+            repository.getServerConfig(forceRefresh = false)
+
+            assertNotNull(
+                fakeConfigDiskSource.serverConfig,
```
Comment (david-livefront) [inline app/src/test/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryTest.kt:118]: Same here

File: app/src/main/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryImpl.kt [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+    override suspend fun getServerConfig(forceRefresh: Boolean): ServerConfig? {
+        val localConfig = configDiskSource.serverConfig
+        val needsRefresh = localConfig == null || localConfig.let {
+            Instant.ofEpochMilli(it.lastSync).isAfter(
+                clock.instant().plusSeconds(MINIMUM_CONFIG_SYNC_INTERVAL),
+            )
+        }
+
+        if (needsRefresh || forceRefresh) {
+            configService.getConfig().onSuccess { configResponse ->
```
Comment (SaintPatrck) [inline app/src/main/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryImpl.kt:47]: ⛏️ Chain formatting
```
configService
    .getConfig()
    .onSuccess { /* wizardry */ }
```

File: app/src/main/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryImpl.kt [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+            .environmentStateFlow
+            .onEach {
+                getServerConfig(true)
+            }
+            .launchIn(unconfinedScope)
+    }
+
+    override suspend fun getServerConfig(forceRefresh: Boolean): ServerConfig? {
+        val localConfig = configDiskSource.serverConfig
+        val needsRefresh = localConfig == null || localConfig.let {
```
Comment (SaintPatrck) [inline app/src/main/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryImpl.kt:40]: `.let` is unnecessary since `localConfig` is non-null when this evaluation takes place.

File: app/src/main/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryImpl.kt [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+        get() = configDiskSource
+            .serverConfigFlow
+            .stateIn(
+                scope = unconfinedScope,
+                started = SharingStarted.Eagerly,
+                initialValue = configDiskSource.serverConfig,
+            )
+
+    companion object {
+        private const val MINIMUM_CONFIG_SYNC_INTERVAL: Long = 60 * 60
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/platform/repository/ServerConfigRepositoryImpl.kt:72]: Can you make this `MINIMUM_CONFIG_SYNC_INTERVAL_SEC` to indicate what unit it is

**Review by SaintPatrck — APPROVED** [review]

**Review by david-livefront — APPROVED** [review]
